File: src/test/java/com/hedera/hashgraph/sdk/NodeCreateTransactionTest.java
// SPDX-License-Identifier: Apache-2.0
package com.hedera.hashgraph.sdk;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.google.protobuf.ByteString;
import com.hedera.hashgraph.sdk.proto.NodeCreateTransactionBody;
import com.hedera.hashgraph.sdk.proto.SchedulableTransactionBody;
import com.hedera.hashgraph.sdk.proto.TransactionBody;
import io.github.jsonSnapshot.SnapshotMatcher;
import java.time.Instant;
import java.util.Arrays;
import java.util.List;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

public class NodeCreateTransactionTest {

    private static final PrivateKey TEST_PRIVATE_KEY = PrivateKey.fromString(
            "302e020100300506032b657004220420db484b828e64b2d8f12ce3c0a0e93a0b8cce7af1bb8f39c97732394482538e10");

    private static final AccountId TEST_ACCOUNT_ID = AccountId.fromString("0.6.9");

    private static final String TEST_DESCRIPTION = "Test description";

    private static final List<Endpoint> TEST_GOSSIP_ENDPOINTS =
            List.of(spawnTestEndpoint((byte) 0), spawnTestEndpoint((byte) 1), spawnTestEndpoint((byte) 2));

    private static final List<Endpoint> TEST_SERVICE_ENDPOINTS = List.of(
            spawnTestEndpoint((byte) 3),
            spawnTestEndpoint((byte) 4),
            spawnTestEndpoint((byte) 5),
            spawnTestEndpoint((byte) 6));

    private static final Endpoint TEST_GRPC_WEB_PROXY_ENDPOINT = spawnTestEndpoint((byte) 3);

    private static final byte[] TEST_GOSSIP_CA_CERTIFICATE = new byte[] {0, 1, 2, 3, 4};

    private static final byte[] TEST_GRPC_CERTIFICATE_HASH = new byte[] {5, 6, 7, 8, 9};

    private static final PublicKey TEST_ADMIN_KEY = PrivateKey.fromString(
                    "302e020100300506032b65700422042062c4b69e9f45a554e5424fb5a6fe5e6ac1f19ead31dc7718c2d980fd1f998d4b")
            .getPublicKey();

    final Instant TEST_VALID_START = Instant.ofEpochSecond(1554158542);

    @BeforeAll
    public static void beforeAll() {
        SnapshotMatcher.start(Snapshot::asJsonString);
    }

    @AfterAll
    public static void afterAll() {
        SnapshotMatcher.validateSnapshots();
    }

    @Test
    void shouldSerialize() {
        SnapshotMatcher.expect(spawnTestTransaction().toString()).toMatchSnapshot();
    }

    private static Endpoint spawnTestEndpoint(byte offset) {
        // Valid endpoint: use domainName only to comply with SDK validation
        return new Endpoint().setDomainName(offset + "unit.test.com").setPort(42 + offset);
    }

    private NodeCreateTransaction spawnTestTransaction() {
        return new NodeCreateTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), TEST_VALID_START))
                .setAccountId(TEST_ACCOUNT_ID)
                .setDescription(TEST_DESCRIPTION)
                .setGossipEndpoints(TEST_GOSSIP_ENDPOINTS)
                .setServiceEndpoints(TEST_SERVICE_ENDPOINTS)
                .setGossipCaCertificate(TEST_GOSSIP_CA_CERTIFICATE)
                .setGrpcCertificateHash(TEST_GRPC_CERTIFICATE_HASH)
                .setAdminKey(TEST_ADMIN_KEY)
                .setMaxTransactionFee(new Hbar(1))
                .setDeclineReward(false) // accepts the reward
                .setGrpcWebProxyEndpoint(TEST_GRPC_WEB_PROXY_ENDPOINT)
                .freeze()
                .sign(TEST_PRIVATE_KEY);
    }

    @Test
    void shouldBytes() throws Exception {
        var tx = spawnTestTransaction();
        var tx2 = NodeCreateTransaction.fromBytes(tx.toBytes());
        assertThat(tx2.toString()).isEqualTo(tx.toString());
    }

    @Test
    void fromScheduledTransaction() {
        var transactionBody = SchedulableTransactionBody.newBuilder()
                .setNodeCreate(NodeCreateTransactionBody.newBuilder().build())
                .build();

        var tx = Transaction.fromScheduledTransaction(transactionBody);

        assertThat(tx).isInstanceOf(NodeCreateTransaction.class);
    }

    @Test
    void shouldBytesNoSetters() throws Exception {
        var tx = new NodeCreateTransaction();
        var tx2 = Transaction.fromBytes(tx.toBytes());
        assertThat(tx2.toString()).isEqualTo(tx.toString());
    }

    @Test
    void testUnrecognizedServicePort() throws Exception {
        var tx = new NodeCreateTransaction()
                .setServiceEndpoints(
                        List.of(new Endpoint().setDomainName("unit.test.com").setPort(50111)));
        var tx2 = NodeCreateTransaction.fromBytes(tx.toBytes());
        assertThat(tx2.toString()).isEqualTo(tx.toString());
    }

    @Test
    void testSetNull() {
        new NodeCreateTransaction()
                .setDescription(null)
                .setAccountId(null)
                .setGossipCaCertificate(null)
                .setGrpcCertificateHash(null)
                .setAdminKey(null);
    }

    @Test
    void setDescriptionRejectsOver100Utf8Bytes() {
        var tx = new NodeCreateTransaction();
        String tooLong = "a".repeat(101);
        assertThrows(IllegalArgumentException.class, () -> tx.setDescription(tooLong));
    }

    @Test
    void setDescriptionAcceptsExactly100Utf8Bytes() {
        var tx = new NodeCreateTransaction();
        String exact = "a".repeat(100);
        tx.setDescription(exact);
        assertThat(tx.getDescription()).isEqualTo(exact);
    }

    @Test
    void setGossipEndpointsRejectsMoreThan10() {
        var tx = new NodeCreateTransaction();
        var endpoints = new java.util.ArrayList<Endpoint>();
        for (int i = 0; i < 11; i++) {
            endpoints.add(new Endpoint().setDomainName("gossip" + i + ".test").setPort(5000 + i));
        }
        assertThrows(IllegalArgumentException.class, () -> tx.setGossipEndpoints(endpoints));
    }

    @Test
    void addGossipEndpointRejectsMoreThan10() {
        var tx = new NodeCreateTransaction();
        for (int i = 0; i < 10; i++) {
            tx.addGossipEndpoint(
                    new Endpoint().setDomainName("gossip" + i + ".test").setPort(5000 + i));
        }
        assertThrows(
                IllegalArgumentException.class,
                () -> tx.addGossipEndpoint(
                        new Endpoint().setDomainName("gossipX.test").setPort(6000)));
    }

    @Test
    void setGossipEndpointsRejectsIpAndDomainTogether() {
        var tx = new NodeCreateTransaction();
        var invalid = new Endpoint()
                .setAddress(new byte[] {1, 2, 3, 4})
                .setDomainName("both.test")
                .setPort(5000);
        assertThrows(IllegalArgumentException.class, () -> tx.setGossipEndpoints(List.of(invalid)));
    }

    @Test
    void setServiceEndpointsRejectsMoreThan8() {
        var tx = new NodeCreateTransaction();
        var endpoints = new java.util.ArrayList<Endpoint>();
        for (int i = 0; i < 9; i++) {
            endpoints.add(new Endpoint().setDomainName("svc" + i + ".test").setPort(6000 + i));
        }
        assertThrows(IllegalArgumentException.class, () -> tx.setServiceEndpoints(endpoints));
    }

    @Test
    void addServiceEndpointRejectsMoreThan8() {
        var tx = new NodeCreateTransaction();
        for (int i = 0; i < 8; i++) {
            tx.addServiceEndpoint(
                    new Endpoint().setDomainName("svc" + i + ".test").setPort(7000 + i));
        }
        assertThrows(
                IllegalArgumentException.class,
                () -> tx.addServiceEndpoint(
                        new Endpoint().setDomainName("svcX.test").setPort(8000)));
    }

    @Test
    void setServiceEndpointsRejectsIpAndDomainTogether() {
        var tx = new NodeCreateTransaction();
        var invalid = new Endpoint()
                .setAddress(new byte[] {5, 6, 7, 8})
                .setDomainName("both.test")
                .setPort(6000);
        assertThrows(IllegalArgumentException.class, () -> tx.setServiceEndpoints(List.of(invalid)));
    }

    @Test
    void setGossipCaCertificateRejectsEmpty() {
        var tx = new NodeCreateTransaction();
        assertThrows(IllegalArgumentException.class, () -> tx.setGossipCaCertificate(new byte[] {}));
    }

    @Test
    void buildRewritesGossipFqdnWithServiceIpFallback() {
        byte[] serviceIp = new byte[] {10, 0, 0, 1};
        Endpoint gossipFqdnOnly =
                new Endpoint().setDomainName("fqdn.example.com").setPort(50211);
        Endpoint serviceIpOnly = new Endpoint().setAddress(serviceIp).setPort(50211);

        var tx = new NodeCreateTransaction()
                .setGossipEndpoints(List.of(gossipFqdnOnly))
                .setServiceEndpoints(List.of(serviceIpOnly));

        var body = tx.build().build();
        var rewritten = body.getGossipEndpoint(0);

        // gossip endpoint should now carry IP and no domain
        org.assertj.core.api.Assertions.assertThat(rewritten.getIpAddressV4().toByteArray())
                .isEqualTo(serviceIp);
        org.assertj.core.api.Assertions.assertThat(rewritten.getDomainName()).isEmpty();
        org.assertj.core.api.Assertions.assertThat(rewritten.getPort()).isEqualTo(50211);
    }

    @Test
    void buildDoesNotRewriteGossipWhenIpPresent() {
        byte[] originalIp = new byte[] {127, 0, 0, 1};
        Endpoint gossipIpOnly = new Endpoint().setAddress(originalIp).setPort(50212);
        byte[] serviceIp = new byte[] {10, 0, 0, 2};
        Endpoint serviceIpOnly = new Endpoint().setAddress(serviceIp).setPort(50211);

        var tx = new NodeCreateTransaction()
                .setGossipEndpoints(List.of(gossipIpOnly))
                .setServiceEndpoints(List.of(serviceIpOnly));

        var body = tx.build().build();
        var ge = body.getGossipEndpoint(0);
        org.assertj.core.api.Assertions.assertThat(ge.getIpAddressV4().toByteArray())
                .isEqualTo(originalIp);
        org.assertj.core.api.Assertions.assertThat(ge.getPort()).isEqualTo(50212);
    }

    @Test
    void buildDoesNotRewriteWhenNoServiceIpAvailable() {
        Endpoint gossipFqdnOnly =
                new Endpoint().setDomainName("fqdn.example.com").setPort(50213);
        Endpoint serviceFqdnOnly =
                new Endpoint().setDomainName("svc.example.com").setPort(50211);

        var tx = new NodeCreateTransaction()
                .setGossipEndpoints(List.of(gossipFqdnOnly))
                .setServiceEndpoints(List.of(serviceFqdnOnly));

        var body = tx.build().build();
        var ge = body.getGossipEndpoint(0);
        org.assertj.core.api.Assertions.assertThat(ge.getIpAddressV4().isEmpty())
                .isTrue();
        org.assertj.core.api.Assertions.assertThat(ge.getDomainName()).isEqualTo("fqdn.example.com");
        org.assertj.core.api.Assertions.assertThat(ge.getPort()).isEqualTo(50213);
    }

    @Test
    void constructNodeCreateTransactionFromTransactionBodyProtobuf() {
        var transactionBodyBuilder = NodeCreateTransactionBody.newBuilder();

        transactionBodyBuilder.setAccountId(TEST_ACCOUNT_ID.toProtobuf());
        transactionBodyBuilder.setDescription(TEST_DESCRIPTION);

        for (Endpoint gossipEndpoint : TEST_GOSSIP_ENDPOINTS) {
            transactionBodyBuilder.addGossipEndpoint(gossipEndpoint.toProtobuf());
        }

        for (Endpoint serviceEndpoint : TEST_SERVICE_ENDPOINTS) {
            transactionBodyBuilder.addServiceEndpoint(serviceEndpoint.toProtobuf());
        }

        transactionBodyBuilder.setGossipCaCertificate(ByteString.copyFrom(TEST_GOSSIP_CA_CERTIFICATE));
        transactionBodyBuilder.setGrpcCertificateHash(ByteString.copyFrom(TEST_GRPC_CERTIFICATE_HASH));
        transactionBodyBuilder.setAdminKey(TEST_ADMIN_KEY.toProtobufKey());

        transactionBodyBuilder.setDeclineReward(true);

        var tx = TransactionBody.newBuilder()
                .setNodeCreate(transactionBodyBuilder.build())
                .build();
        var nodeCreateTransaction = new NodeCreateTransaction(tx);

        assertThat(nodeCreateTransaction.getAccountId()).isEqualTo(TEST_ACCOUNT_ID);
        assertThat(nodeCreateTransaction.getDescription()).isEqualTo(TEST_DESCRIPTION);
        assertThat(nodeCreateTransaction.getGossipEndpoints()).hasSize(TEST_GOSSIP_ENDPOINTS.size());
        assertThat(nodeCreateTransaction.getServiceEndpoints()).hasSize(TEST_SERVICE_ENDPOINTS.size());
        assertThat(nodeCreateTransaction.getGossipCaCertificate()).isEqualTo(TEST_GOSSIP_CA_CERTIFICATE);
        assertThat(nodeCreateTransaction.getGrpcCertificateHash()).isEqualTo(TEST_GRPC_CERTIFICATE_HASH);
        assertThat(nodeCreateTransaction.getAdminKey()).isEqualTo(TEST_ADMIN_KEY);
    }

    @Test
    void getSetAccountId() {
        var nodeCreateTransaction = new NodeCreateTransaction().setAccountId(TEST_ACCOUNT_ID);
        assertThat(nodeCreateTransaction.getAccountId()).isEqualTo(TEST_ACCOUNT_ID);
    }

    @Test
    void getSetAccountIdFrozen() {
        var tx = spawnTestTransaction();
        assertThrows(IllegalStateException.class, () -> tx.setAccountId(TEST_ACCOUNT_ID));
    }

    @Test
    void getSetDescription() {
        var nodeCreateTransaction = new NodeCreateTransaction().setDescription(TEST_DESCRIPTION);
        assertThat(nodeCreateTransaction.getDescription()).isEqualTo(TEST_DESCRIPTION);
    }

    @Test
    void getSetDescriptionFrozen() {
        var tx = spawnTestTransaction();
        assertThrows(IllegalStateException.class, () -> tx.setDescription(TEST_DESCRIPTION));
    }

    @Test
    void getSetGossipEndpoints() {
        var nodeCreateTransaction = new NodeCreateTransaction().setGossipEndpoints(TEST_GOSSIP_ENDPOINTS);
        assertThat(nodeCreateTransaction.getGossipEndpoints()).isEqualTo(TEST_GOSSIP_ENDPOINTS);
    }

    @Test
    void setTestGossipEndpointsFrozen() {
        var tx = spawnTestTransaction();
        assertThrows(IllegalStateException.class, () -> tx.setGossipEndpoints(TEST_GOSSIP_ENDPOINTS));
    }

    @Test
    void getSetServiceEndpoints() {
        var nodeCreateTransaction = new NodeCreateTransaction().setServiceEndpoints(TEST_SERVICE_ENDPOINTS);
        assertThat(nodeCreateTransaction.getServiceEndpoints()).isEqualTo(TEST_SERVICE_ENDPOINTS);
    }

    @Test
    void getSetServiceEndpointsFrozen() {
        var tx = spawnTestTransaction();
        assertThrows(IllegalStateException.class, () -> tx.setServiceEndpoints(TEST_SERVICE_ENDPOINTS));
    }

    @Test
    void getSetGossipCaCertificate() {
        var nodeCreateTransaction = new NodeCreateTransaction().setGossipCaCertificate(TEST_GOSSIP_CA_CERTIFICATE);
        assertThat(nodeCreateTransaction.getGossipCaCertificate()).isEqualTo(TEST_GOSSIP_CA_CERTIFICATE);
    }

    @Test
    void getSetGossipCaCertificateFrozen() {
        var tx = spawnTestTransaction();
        assertThrows(IllegalStateException.class, () -> tx.setGossipCaCertificate(TEST_GOSSIP_CA_CERTIFICATE));
    }

    @Test
    void getSetGrpcCertificateHash() {
        var nodeCreateTransaction = new NodeCreateTransaction().setGrpcCertificateHash(TEST_GRPC_CERTIFICATE_HASH);
        assertThat(nodeCreateTransaction.getGrpcCertificateHash()).isEqualTo(TEST_GRPC_CERTIFICATE_HASH);
    }

    @Test
    void getSetGrpcCertificateHashFrozen() {
        var tx = spawnTestTransaction();
        assertThrows(IllegalStateException.class, () -> tx.setGrpcCertificateHash(TEST_GRPC_CERTIFICATE_HASH));
    }

    @Test
    void getSetAdminKey() {
        var nodeCreateTransaction = new NodeCreateTransaction().setAdminKey(TEST_ADMIN_KEY);
        assertThat(nodeCreateTransaction.getAdminKey()).isEqualTo(TEST_ADMIN_KEY);
    }

    @Test
    void getSetAdminKeyFrozen() {
        var tx = spawnTestTransaction();
        assertThrows(IllegalStateException.class, () -> tx.setAdminKey(TEST_ADMIN_KEY));
    }

    @Test
    void getSetDeclineReward() {
        var nodeCreateTransaction = new NodeCreateTransaction().setDeclineReward(true);
        assertThat(nodeCreateTransaction.getDeclineReward()).isTrue();
    }

    @Test
    void getSetDeclineRewardFrozen() {
        var tx = spawnTestTransaction();
        assertThrows(IllegalStateException.class, () -> tx.setDeclineReward(false));
    }

    @Test
    void getGrpcWebProxyEndpoint() {
        var nodeCreateTransaction = new NodeCreateTransaction().setGrpcWebProxyEndpoint(TEST_GRPC_WEB_PROXY_ENDPOINT);
        assertThat(nodeCreateTransaction.getGrpcWebProxyEndpoint()).isEqualTo(TEST_GRPC_WEB_PROXY_ENDPOINT);
    }

    @Test
    void setGrpcWebProxyEndpointRequiresFrozen() {
        var tx = spawnTestTransaction();
        assertThrows(IllegalStateException.class, () -> tx.setGrpcWebProxyEndpoint(TEST_GRPC_WEB_PROXY_ENDPOINT));
    }
}

File: src/Hedera.Hashgraph.Tests/SDK/Networking/NodeCreateTransactionTest.cs
// SPDX-License-Identifier: Apache-2.0
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;

using Hedera.Hashgraph.SDK;
using Hedera.Hashgraph.SDK.Account;
using Hedera.Hashgraph.SDK.HBar;
using Hedera.Hashgraph.SDK.Keys;
using Hedera.Hashgraph.SDK.Networking;
using Hedera.Hashgraph.SDK.Transactions;

using System;
using System.Collections.Generic;
using System.Linq;

using VerifyXunit;

namespace Hedera.Hashgraph.Tests.SDK.Networking
{
    public class NodeCreateTransactionTest
    {
        private static readonly PrivateKey TEST_PRIVATE_KEY = PrivateKey.FromString("302e020100300506032b657004220420db484b828e64b2d8f12ce3c0a0e93a0b8cce7af1bb8f39c97732394482538e10");
        private static readonly AccountId TEST_ACCOUNT_ID = AccountId.FromString("0.6.9");
        private static readonly string TEST_DESCRIPTION = "Test description";
        private static readonly List<Endpoint> TEST_GOSSIP_ENDPOINTS = [ SpawnTestEndpoint((byte)0), SpawnTestEndpoint((byte)1), SpawnTestEndpoint((byte)2) ];
        private static readonly List<Endpoint> TEST_SERVICE_ENDPOINTS = [ SpawnTestEndpoint((byte)3), SpawnTestEndpoint((byte)4), SpawnTestEndpoint((byte)5), SpawnTestEndpoint((byte)6) ];
        private static readonly Endpoint TEST_GRPC_WEB_PROXY_ENDPOINT = SpawnTestEndpoint((byte)3);
        private static readonly byte[] TEST_GOSSIP_CA_CERTIFICATE = new byte[]
        {
            0,
            1,
            2,
            3,
            4
        };
        private static readonly byte[] TEST_GRPC_CERTIFICATE_HASH = new byte[]
        {
            5,
            6,
            7,
            8,
            9
        };
        private static readonly PublicKey TEST_ADMIN_KEY = PrivateKey.FromString("302e020100300506032b65700422042062c4b69e9f45a554e5424fb5a6fe5e6ac1f19ead31dc7718c2d980fd1f998d4b").GetPublicKey();
        readonly DateTimeOffset TEST_VALID_START = DateTimeOffset.FromUnixTimeMilliseconds(1554158542);

        public virtual void ShouldSerialize()
        {
            Verifier.Verify(SpawnTestTransaction().ToString());
        }

        private static Endpoint SpawnTestEndpoint(byte offset)
        {
            // Valid endpoint: use domainName only to comply with SDK validation
            return new Endpoint
            {
				DomainName = offset + "unit.test.com",
				Port = 42 + offset,
			};
        }

        private NodeCreateTransaction SpawnTestTransaction()
        {
            return new NodeCreateTransaction
            {
				AccountId = TEST_ACCOUNT_ID,
				Description = TEST_DESCRIPTION,
				GossipEndpoints = TEST_GOSSIP_ENDPOINTS,
				ServiceEndpoints = TEST_SERVICE_ENDPOINTS,
				GossipCaCertificate = TEST_GOSSIP_CA_CERTIFICATE,
				GrpcCertificateHash = TEST_GRPC_CERTIFICATE_HASH,

				AdminKey = TEST_ADMIN_KEY,
				MaxTransactionFee = new Hbar(1),
				DeclineReward = false,
				GrpcWebProxyEndpoint = TEST_GRPC_WEB_PROXY_ENDPOINT,

				NodeAccountIds = [AccountId.FromString("0.0.5005"), AccountId.FromString("0.0.5006")],
				TransactionId = TransactionId.WithValidStart(AccountId.FromString("0.0.5006"), TEST_VALID_START),
			}
            .Freeze()
            .Sign(TEST_PRIVATE_KEY);
        }

        public virtual void ShouldBytes()
        {
            var tx = SpawnTestTransaction();
            var tx2 = Transaction.FromBytes<NodeCreateTransaction>(tx.ToBytes());

            Assert.Equal(tx2.ToString(), tx.ToString());
        }

        public virtual void FromScheduledTransaction()
        {
            var transactionBody = new Proto.SchedulableTransactionBody
            {
				NodeCreate = new Proto.NodeCreateTransactionBody()
			};
            var tx = Transaction.FromScheduledTransaction<NodeCreateTransaction>(transactionBody);

            Assert.IsType<NodeCreateTransaction>(tx);
        }

        public virtual void ShouldBytesNoSetters()
        {
            var tx = new NodeCreateTransaction();
            var tx2 = Transaction.FromBytes<NodeCreateTransaction>(tx.ToBytes());

            Assert.Equal(tx2.ToString(), tx.ToString());
        }

        public virtual void TestUnrecognizedServicePort()
        {
            var tx = new NodeCreateTransaction
            {
				ServiceEndpoints = [ new Endpoint
				{
					DomainName = "unit.test.com",
					Port = 50111,
				} ]
			};
            var tx2 = Transaction.FromBytes<NodeCreateTransaction>(tx.ToBytes());

            Assert.Equal(tx2.ToString(), tx.ToString());
        }

        public virtual void TestSetNull()
        {
            new NodeCreateTransaction
            {
                Description = null,
                AccountId = null,
                GossipCaCertificate = null,
                GrpcCertificateHash = null,
                AdminKey = null,
            };
        }

        public virtual void SetDescriptionRejectsOver100Utf8Bytes()
        {
            var tx = new NodeCreateTransaction();
            string tooLong = string.Join(string.Empty, Enumerable.Repeat("a", 101));
			Assert.Throws<ArgumentException>(() => tx.Description = tooLong);
        }

        public virtual void SetDescriptionAcceptsExactly100Utf8Bytes()
        {
            var tx = new NodeCreateTransaction();
            string exact = string.Join(string.Empty, Enumerable.Repeat("a", 100));
            tx.Description = exact;

            Assert.Equal(tx.Description, exact);
        }

        public virtual void SetGossipEndpointsRejectsMoreThan10()
        {
            var tx = new NodeCreateTransaction();
            var endpoints = new List<Endpoint>();
            for (int i = 0; i < 11; i++)
				endpoints.Add(new Endpoint
				{
					DomainName = "gossip" + i + ".test",
					Port = 5000 + i,
				});


            Assert.Throws<ArgumentException>(() => tx.GossipEndpoints = endpoints);
        }

        public virtual void AddGossipEndpointRejectsMoreThan10()
        {
            var tx = new NodeCreateTransaction();
            
			for (int i = 0; i < 10; i++)
				tx.GossipEndpoints.Add(new Endpoint
				{
					DomainName = "gossip" + i + ".test",
					Port = 5000 + i
				});

			Assert.Throws<ArgumentException>(() => tx.GossipEndpoints.Add(new Endpoint
            {
				DomainName = "gossipX.test",
				Port = 6000
			}));
        }

        public virtual void SetGossipEndpointsRejectsIpAndDomainTogether()
        {
            var tx = new NodeCreateTransaction();
            var invalid = new Endpoint
            {
				Address = new byte[] { 1, 2, 3, 4 },
				DomainName = "both.test",
				Port = 5000,
			};

            Assert.Throws<ArgumentException>(() => tx.GossipEndpoints = [invalid]);
        }

        public virtual void SetServiceEndpointsRejectsMoreThan8()
        {
            var tx = new NodeCreateTransaction();
            var endpoints = new List<Endpoint>();
            for (int i = 0; i < 9; i++)
				endpoints.Add(new Endpoint
				{
					DomainName = "svc" + i + ".test",
					Port = 6000 + i,
				});

			Assert.Throws<ArgumentException>(() => tx.ServiceEndpoints = endpoints);
        }

        public virtual void AddServiceEndpointRejectsMoreThan8()
        {
            var tx = new NodeCreateTransaction();
            for (int i = 0; i < 8; i++)
				tx.ServiceEndpoints.Add(new Endpoint
				{
					DomainName = "svc" + i + ".test",
					Port = 7000 + i,
				});

			Assert.Throws<ArgumentException>(() => tx.ServiceEndpoints.Add(new Endpoint
			{
				DomainName = "svcX.test",
				Port = 8000,
			}));
        }

        public virtual void SetServiceEndpointsRejectsIpAndDomainTogether()
        {
            var tx = new NodeCreateTransaction();
            var invalid = new Endpoint
			{
				Address = new byte[] { 5, 6, 7, 8 },
				DomainName = "both.test",
				Port = 6000
			};

            Assert.Throws<ArgumentException>(() => tx.ServiceEndpoints = [invalid]);
        }

        public virtual void SetGossipCaCertificateRejectsEmpty()
        {
            var tx = new NodeCreateTransaction();

            Assert.Throws<ArgumentException>(() => tx.GossipCaCertificate = new byte[] { });
        }

        public virtual void BuildRewritesGossipFqdnWithServiceIpFallback()
        {
            byte[] serviceIp = new byte[]
            {
                10,
                0,
                0,
                1
            };
            
			Endpoint gossipFqdnOnly = new Endpoint
			{
				DomainName = "fqdn.example.com",
				Port = 50211
			};
			Endpoint serviceFqdnOnly = new Endpoint
			{
				Address = serviceIp,
				Port = 50211
			};
			var tx = new NodeCreateTransaction
			{
				GossipEndpoints = [gossipFqdnOnly],
				ServiceEndpoints = [serviceFqdnOnly],
			};
			var rewritten = tx.GossipEndpoints[0];

			// gossip endpoint should now carry IP and no domain
			Assert.Equal(rewritten.Address, serviceIp);
            Assert.Empty(rewritten.DomainName);
            Assert.Equal(rewritten.Port, 50211);
        }

        public virtual void BuildDoesNotRewriteGossipWhenIpPresent()
        {
            byte[] originalIp = new byte[]
            {
                127,
                0,
                0,
                1
            };
            byte[] serviceIp = new byte[]
            {
                10,
                0,
                0,
                2
            };

			Endpoint gossipIpOnly = new Endpoint
			{
				Address = originalIp,
				Port = 50212
			};
			Endpoint serviceIpOnly = new Endpoint
			{
				Address = serviceIp,
				Port = 50211
			};
			var tx = new NodeCreateTransaction
			{
				GossipEndpoints = [gossipIpOnly],
				ServiceEndpoints = [serviceIpOnly],
			};
			var ge = tx.GossipEndpoints[0];

			Assert.Equal(ge.Address, originalIp);
            Assert.Equal(ge.Port, 50212);
        }

        public virtual void BuildDoesNotRewriteWhenNoServiceIpAvailable()
        {
            Endpoint gossipFqdnOnly = new Endpoint
            {
                DomainName = "fqdn.example.com",
                Port = 50213 
            };
            Endpoint serviceFqdnOnly = new Endpoint
            {
                DomainName = "svc.example.com",
                Port = 50211 
            };
            var tx = new NodeCreateTransaction
            {
				GossipEndpoints = [gossipFqdnOnly],
				ServiceEndpoints = [serviceFqdnOnly],
			};
            var ge = tx.GossipEndpoints[0];

            Assert.True(ge.Address.Length == 0);
            Assert.Equal(ge.DomainName, "fqdn.example.com");
            Assert.Equal(ge.Port, 50213);
        }

        public virtual void ConstructNodeCreateTransactionFromTransactionBodyProtobuf()
        {
            var transactionBodyBuilder = new Proto.NodeCreateTransactionBody
            {
				GossipCaCertificate = ByteString.CopyFrom(TEST_GOSSIP_CA_CERTIFICATE),
				GrpcCertificateHash = ByteString.CopyFrom(TEST_GRPC_CERTIFICATE_HASH),
				AdminKey = TEST_ADMIN_KEY.ToProtobufKey(),
				DeclineReward = true,
				AccountId = TEST_ACCOUNT_ID.ToProtobuf(),
				Description = TEST_DESCRIPTION,
			};

            foreach (Endpoint gossipEndpoint in TEST_GOSSIP_ENDPOINTS)
				transactionBodyBuilder.GossipEndpoint.Add(gossipEndpoint.ToProtobuf());

			foreach (Endpoint serviceEndpoint in TEST_SERVICE_ENDPOINTS)
				transactionBodyBuilder.ServiceEndpoint.Add(serviceEndpoint.ToProtobuf());



			var nodeCreateTransaction = new NodeCreateTransaction(new Proto.TransactionBody
			{
				NodeCreate = transactionBodyBuilder
			});
            Assert.Equal(nodeCreateTransaction.AccountId, TEST_ACCOUNT_ID);
            Assert.Equal(nodeCreateTransaction.Description, TEST_DESCRIPTION);
            Assert.Equal(nodeCreateTransaction.GossipEndpoints.Count, TEST_GOSSIP_ENDPOINTS.Count);
            Assert.Equal(nodeCreateTransaction.ServiceEndpoints.Count, TEST_SERVICE_ENDPOINTS.Count);
            Assert.Equal(nodeCreateTransaction.GossipCaCertificate, TEST_GOSSIP_CA_CERTIFICATE);
            Assert.Equal(nodeCreateTransaction.GrpcCertificateHash, TEST_GRPC_CERTIFICATE_HASH);
            Assert.Equal(nodeCreateTransaction.AdminKey, TEST_ADMIN_KEY);
        }

		public virtual void GetSetAccountId()
		{
			var nodeCreateTransaction = new NodeCreateTransaction
			{
				AccountId = TEST_ACCOUNT_ID
			};
			Assert.Equal(nodeCreateTransaction.AccountId, TEST_ACCOUNT_ID);
		}

		public virtual void GetSetAccountIdFrozen()
		{
			var tx = SpawnTestTransaction();
			Assert.Throws<InvalidOperationException>(() => tx.AccountId = TEST_ACCOUNT_ID);
		}

		public virtual void GetSetDescription()
		{
			var nodeCreateTransaction = new NodeCreateTransaction
			{
				Description = TEST_DESCRIPTION
			};
			Assert.Equal(nodeCreateTransaction.Description, TEST_DESCRIPTION);
		}

		public virtual void GetSetDescriptionFrozen()
		{
			var tx = SpawnTestTransaction();
			Assert.Throws<InvalidOperationException>(() => tx.Description = TEST_DESCRIPTION);
		}

		public virtual void GetSetGossipEndpoints()
		{
			var nodeCreateTransaction = new NodeCreateTransaction
			{
				GossipEndpoints = TEST_GOSSIP_ENDPOINTS
			};
			Assert.Equal(nodeCreateTransaction.GossipEndpoints, TEST_GOSSIP_ENDPOINTS);
		}

		public virtual void SetTestGossipEndpointsFrozen()
		{
			var tx = SpawnTestTransaction();
			Assert.Throws<InvalidOperationException>(() => tx.GossipEndpoints = TEST_GOSSIP_ENDPOINTS);
		}

		public virtual void GetSetServiceEndpoints()
		{
			var nodeCreateTransaction = new NodeCreateTransaction
			{
				ServiceEndpoints = TEST_SERVICE_ENDPOINTS
			};
			Assert.Equal(nodeCreateTransaction.ServiceEndpoints, TEST_SERVICE_ENDPOINTS);
		}

		public virtual void GetSetServiceEndpointsFrozen()
		{
			var tx = SpawnTestTransaction();
			Assert.Throws<InvalidOperationException>(() => tx.ServiceEndpoints = TEST_SERVICE_ENDPOINTS);
		}

		public virtual void GetSetGossipCaCertificate()
		{
			var nodeCreateTransaction = new NodeCreateTransaction
			{
				GossipCaCertificate = TEST_GOSSIP_CA_CERTIFICATE
			};
			Assert.Equal(nodeCreateTransaction.GossipCaCertificate, TEST_GOSSIP_CA_CERTIFICATE);
		}

		public virtual void GetSetGossipCaCertificateFrozen()
		{
			var tx = SpawnTestTransaction();
			Assert.Throws<InvalidOperationException>(() => tx.GossipCaCertificate = TEST_GOSSIP_CA_CERTIFICATE);
		}

		public virtual void GetSetGrpcCertificateHash()
		{
			var nodeCreateTransaction = new NodeCreateTransaction
			{
				GrpcCertificateHash = TEST_GRPC_CERTIFICATE_HASH
			};
			Assert.Equal(nodeCreateTransaction.GrpcCertificateHash, TEST_GRPC_CERTIFICATE_HASH);
		}

		public virtual void GetSetGrpcCertificateHashFrozen()
		{
			var tx = SpawnTestTransaction();
			Assert.Throws<InvalidOperationException>(() => tx.GrpcCertificateHash = TEST_GRPC_CERTIFICATE_HASH);
		}

		public virtual void GetSetAdminKey()
		{
			var nodeCreateTransaction = new NodeCreateTransaction
			{
				AdminKey = TEST_ADMIN_KEY
			};
			Assert.Equal(nodeCreateTransaction.AdminKey, TEST_ADMIN_KEY);
		}

		public virtual void GetSetAdminKeyFrozen()
		{
			var tx = SpawnTestTransaction();
			Assert.Throws<InvalidOperationException>(() => tx.AdminKey = TEST_ADMIN_KEY);
		}

		public virtual void GetSetDeclineReward()
		{
			var nodeCreateTransaction = new NodeCreateTransaction
			{
				DeclineReward = true
			};
			Assert.True(nodeCreateTransaction.DeclineReward);
		}

		public virtual void GetSetDeclineRewardFrozen()
		{
			var tx = SpawnTestTransaction();
			Assert.Throws<InvalidOperationException>(() => tx.DeclineReward = false);
		}

		public virtual void GetGrpcWebProxyEndpoint()
		{
			var nodeCreateTransaction = new NodeCreateTransaction
			{
				GrpcWebProxyEndpoint = TEST_GRPC_WEB_PROXY_ENDPOINT
			};
			Assert.Equal(nodeCreateTransaction.GrpcWebProxyEndpoint, TEST_GRPC_WEB_PROXY_ENDPOINT);
		}

		public virtual void SetGrpcWebProxyEndpointRequiresFrozen()
		{
			var tx = SpawnTestTransaction();
			Assert.Throws<InvalidOperationException>(() => tx.GrpcWebProxyEndpoint = TEST_GRPC_WEB_PROXY_ENDPOINT);
		}
	}
}